### ISSUE TITLE
[no-unnecessary-condition] false positive with change in function call

### ISSUE THREAD

**Issue by JounQin** [typescript-eslint/typescript-eslint#3732]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/no-unnecessary-condition": [2]
  }
}
```

```TS
export const test = () => {
  let val = false

  const fn = () => {
    val = true
  }

  fn()

  if (val) { // error reported here
    console.log('')
  }
}
```

```json
{
  "extends": "@1stg/tsconfig",
  "compilerOptions": {
    "outDir": "lib",
    "target": "ES2019"
  }
}
```

**Expected Result**

No error

**Actual Result**

```log
  10:7  error  Unnecessary conditional, value is always falsy  @typescript-eslint/no-unnecessary-condition
```

**Additional Info**

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.29.1` |
| `@typescript-eslint/parser`        | `4.29.1` |
| `TypeScript`                       | `4.3.5` |
| `ESLint`                           | `7.32.0` |
| `node`                             | `v12.22.1` |

**Comment by bradzacher** [comment]
Typescript's flow analysis does not cover this. Which means it reports the type of `val` as `false`.

Nothing we can do about this as we can only act on the TS types.

**Comment by JounQin** [comment]
OK, I raised https://github.com/microsoft/TypeScript/issues/45443 for tracking.
